librz/arch/isa/hexagon/hexagon_arch.c

```diff
@@ -1121,6 +1130,83 @@ RZ_IPI void hexagon_pkt_mark_tail_calls(HexPkt *pkt) {
 	hic->ana_op.type = RZ_ANALYSIS_OP_TYPE_TAIL | RZ_ANALYSIS_OP_TYPE_RET;
 }
 
+static RZ_BORROW HexInsnContainer *decode_hic(HexState *state, HexReversedOpcode *rz_reverse, RZ_BORROW RzBuffer *buffer, const ut64 addr) {
+	ut8 tmp[HEX_INSN_SIZE] = { 0 };
+	ut32 bytes = rz_buf_read(buffer, tmp, 4);
+	if (bytes != HEX_INSN_SIZE) {
+		RZ_LOG_ERROR("Failed to read from buffer!\n");
+		return NULL;
+	}
+	ut32 data = rz_read_le32(tmp);
+	ut8 parse_bits = HEX_PARSE_BITS_FROM_UT32(data);
+	HexInsnContainer hic_new = { 0 };
+	setup_new_hic(&hic_new, rz_reverse, addr, parse_bits, data);
+
+	// Add to state as not yet fully decoded packet.
+	HexInsnContainer *hic = hex_add_hic_to_state(state, &hic_new);
+	if (!hic) {
+		RZ_LOG_ERROR("Could not add incsturction container to state.\n");
+		return NULL;
+	}
+	HexPkt *p = hex_get_pkt(state, hic->addr);
+
+	// Do disassembly and analysis
+	hexagon_disasm_instruction(state, data, hic, p);
+	return hic;
+}
+
+/**
+ * \brief Returns the address at which the decoding must start to get a valid packet at \p addr.
+ * The \p buffer seek is set to the position to start reading from.
+ *
+ * \return The address to start decoding. It always returns an address <= \p addr
+ * and with an offset with an multiple of HEX_INSN_SIZE.
+ */
+static ut64 get_pre_decoding_start(RZ_BORROW RzBuffer *buffer, ut64 addr) {
+	rz_return_val_if_fail(buffer, addr);
+	
+	ut64 seek = rz_buf_tell(buffer);
+	if (addr < HEX_INSN_SIZE && seek < HEX_INSN_SIZE) {
+		return addr;
+	}
+
+	size_t look_back = 0;
+	bool is_last_insn = false;
+	// Search until we cross a boundary or have found a last instruction.
+	while (addr >= HEX_INSN_SIZE && seek >= HEX_INSN_SIZE && look_back < 4 && !is_last_insn) {
+		ut8 tmp[HEX_INSN_SIZE] = { 0 };
+		ut32 bytes = rz_buf_read(buffer, tmp, 4);
+		if (bytes != HEX_INSN_SIZE) {
+			return addr;
+		}
+		ut32 data = rz_read_le32(tmp);
+		is_last_insn = is_last_instr(HEX_PARSE_BITS_FROM_UT32(data));
+		if (!rz_buf_seek(buffer, -HEX_INSN_SIZE, RZ_BUF_CUR)) {
+			return addr;
+		}
+		addr -= HEX_INSN_SIZE;
+		look_back++;
+	}
+	return addr;
+}
+
+static bool perform_hacks(HexState **state, RzBuffer **buffer, RzAsm **rz_asm, RzAnalysis **rz_analysis) {
+	if (*rz_analysis) {
+		*rz_asm = rz_analysis_to_rz_asm(*rz_analysis);
+		assert(*rz_asm && (*rz_asm)->cur && (*rz_analysis)->cur && RZ_STR_EQ((*rz_asm)->cur->arch, (*rz_analysis)->cur->arch));
+	} else if (*rz_asm) {
+		*rz_analysis = rz_asm_to_rz_analysis(rz_asm);
+		assert(*rz_analysis && (*rz_asm)->cur && (*rz_analysis)->cur && RZ_STR_EQ((*rz_asm)->cur->arch, (*rz_analysis)->cur->arch));
+	} else {
+		assert(0 && "Requires either RzAsm or RzAnalysis");
+	}
+	*state = (*rz_asm)->plugin_data;
+	rz_return_val_if_fail(*state, false);
+	*buffer = rz_buf_new_with_io(&(*rz_analysis)->iob);
+	rz_return_val_if_fail(*buffer, false);
+	return true;
+}
+
```

Comment:
@wargio This would be a possible way to initialize a buffer and get the `plugin_data` no matter if `RzAsm` or `RzAnalysis` was passed.

Reply:
Alhtough it still needs a version which handles the `Bytes` case of the buffer.